crates/moss-cache/src/lib.rs

```diff
@@ -0,0 +1,250 @@
+}
+
+impl MossCache {
+    pub fn contains(&self, key: &str) -> bool {
+        /// Check if there's stored cache value for key
+        self.cache.contains_key(key)
+    }
+    pub fn insert<T>(&self, key: &str, value: T, ttl: Duration)
+    where
+        T: Any + Send + Sync,
+    {
+        /// Insert a key-value pair into the cache with a given TTL
+        self.cache
+            .insert(key.to_string(), MossCacheItem::new(Arc::new(value), ttl));
+    }
+
+    pub fn delete(&self, key: &str) {
+        /// Deletes the stored value for key
+        self.cache.remove(key);
+    }
+
+    pub fn get<T>(&self, key: &str) -> Result<Arc<T>, MossCacheError>
+    where
+        T: Any + Send + Sync,
+    {
+        /// Attempt to first get the value of the key; if not, it will first update the cache value
+        /// using the registered update function
+        match self.cache.get(key) {
+            Some(cache_item) => cache_item.get::<T>(key),
+            None => {
+                let retriever = self
+                    .retrievers
+                    .get(key)
+                    .ok_or(MossCacheError::NonexistentKey {
+                        key: key.to_string(),
+                    })?
+                    .value()
+                    .clone();
+                self.cache
+                    .get_with(key.to_string(), move || retriever())
+                    .get::<T>(key)
+            }
+        }
+    }
+
+    pub fn take<T>(&self, key: &str) -> Result<T, MossCacheError>
+    where
+        T: Any + Clone + Send + Sync,
+    {
+        /// Removes the stored value for key and get the raw T value
+        /// To get T from Arc<T>, T must be Clone
+        let cache_item = self
+            .cache
+            .remove(key)
+            .ok_or(MossCacheError::NonexistentKey {
+                key: key.to_string(),
+            })?;
+        cache_item
+            .get::<T>(key)
+            .map(|arc_t| Arc::<T>::unwrap_or_clone(arc_t))
+    }
+
+    pub fn register<T>(&self, key: &str, ttl: Duration, update: impl Fn() -> T + 'static) -> Arc<T>
+    where
+        T: Any + Send + Sync,
+    {
+        /// Register a key with an update function, which will be called when the entry is expired
+        let update = Arc::new(update);
+        let retriever = {
+            let update = Arc::clone(&update);
+            Arc::new(move || MossCacheItem::new(Arc::new(update()), ttl))
+        };
+        self.retrievers.insert(key.to_string(), retriever.clone());
+        self.cache
+            .get_with(key.to_string(), move || retriever())
+            .get::<T>(key)
+            .unwrap()
+    }
+
+    pub fn deregister(&self, key: &str) -> Option<()> {
+        /// Delete cached value and the registered function
+        /// If no function has been registered for this key, returns None
+        /// Otherwise returns Some(())
+        self.cache.invalidate(key);
+        self.retrievers.remove(key).map(|_| ())
```

Comment:
It might be better to add more clarity here by creating an error if an attempt is made to deregister a listener that no longer exists. This would prevent erroneous implementations on the caller’s side, where multiple deregistrations of a non-existent entity occur, but we remain unaware that we’re performing meaningless operations.

Reply:
Good idea!